// @uni-helper/vite-plugin-uni-platform@0.0.5/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const node_fs = require('node:fs');
const node_path = require('node:path');
const uniEnv = require('@uni-helper/uni-env');
const vite = require('vite');
const node_module = require('node:module');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
const platform = uniEnv.customScript || uniEnv.platform;
const _require = typeof require === "function" ? require : node_module.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
const uniUtils = _require("@dcloudio/uni-cli-shared/dist/utils.js");
const uniResolve = _require("@dcloudio/uni-cli-shared/dist/resolve.js");
const constants = _require("@dcloudio/uni-cli-shared/dist/constants.js");
uniUtils.normalizePagePath = function(pagePath, platform2) {
  const absolutePagePath = node_path.resolve(uniEnv.inputDir, pagePath);
  let extensions = constants.PAGE_EXTNAME;
  if (uniEnv.isApp)
    extensions = constants.PAGE_EXTNAME_APP;
  for (let i = 0; i < extensions.length; i++) {
    const extname = extensions[i];
    if (node_fs.existsSync(absolutePagePath + extname))
      return pagePath + extname;
    const withPlatform = `${absolutePagePath}.${platform2}${extname}`;
    if (node_fs.existsSync(withPlatform))
      return pagePath + extname;
  }
  console.error(`${pagePath} not found`);
};
const requireResolve = uniResolve.requireResolve;
uniResolve.requireResolve = function(filename, basedir) {
  try {
    return requireResolve(filename, basedir);
  } catch (error) {
    const { ext, base, dir } = node_path.parse(filename);
    filename = `${dir}/${base}.${platform}${ext ? `.${ext}` : ""}`;
    return requireResolve(filename, basedir);
  }
};

function resolveOptions(userOptions) {
  return {
    include: "**/*",
    exclude: [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
    ...userOptions
  };
}
function hasPlatformSuffix(id) {
  const fileName = node_path.basename(id, node_path.extname(id));
  return new RegExp(`\\.${platform}$`).test(fileName);
}
function joinPlatformSuffix(id) {
  return id.replace(/(.*)\.(.*)$/, `$1.${platform}.$2`);
}
function VitePluginUniPlatform(userOptions = {}) {
  const options = resolveOptions(userOptions);
  const filter = vite.createFilter(options.include, options.exclude);
  return {
    name: "vite-plugin-uni-platform",
    enforce: "pre",
    async resolveId(source, importer, options2) {
      if (hasPlatformSuffix(source))
        return null;
      const sourceResolution = await this.resolve(source, importer, {
        ...options2,
        skipSelf: true
        // 避免无限循环
      });
      if (sourceResolution)
        return null;
      const platformSource = joinPlatformSuffix(source);
      const resolution = await this.resolve(platformSource, importer, { ...options2, skipSelf: true });
      if (!resolution || resolution.external)
        return resolution;
      if (!filter(resolution.id))
        return null;
      const sourceId = vite.normalizePath(node_path.resolve(node_path.dirname(importer), source));
      const isVue = resolution.id.endsWith("vue");
      return uniEnv.isMp && isVue ? sourceId : resolution;
    },
    // 自定义加载器，尝试将所有不带 {platform} 后缀的文件拼接 {platform} 后去加载
    async load(id) {
      let platformId = id;
      if (!hasPlatformSuffix(id))
        platformId = joinPlatformSuffix(id);
      if (!filter(platformId))
        return;
      if (platformId && platformId !== id && node_fs.existsSync(platformId)) {
        return node_fs.readFileSync(platformId, {
          encoding: "utf-8"
        });
      }
    }
  };
}

exports.VitePluginUniPlatform = VitePluginUniPlatform;
exports.default = VitePluginUniPlatform;
