// openapi-ts-request@1.10.0/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.displayReactQueryMode = exports.ReactQueryMode = exports.PriorityRule = exports.SchemaObjectType = exports.SchemaObjectFormat = void 0;
var SchemaObjectFormat;
(function (SchemaObjectFormat) {
    SchemaObjectFormat["int32"] = "int32";
    SchemaObjectFormat["int64"] = "int64";
    SchemaObjectFormat["float"] = "float";
    SchemaObjectFormat["double"] = "double";
    SchemaObjectFormat["byte"] = "byte";
    SchemaObjectFormat["binary"] = "binary";
    SchemaObjectFormat["date"] = "date";
    SchemaObjectFormat["dateTime"] = "date-time";
    SchemaObjectFormat["password"] = "password";
    SchemaObjectFormat["base64"] = "base64";
})(SchemaObjectFormat || (exports.SchemaObjectFormat = SchemaObjectFormat = {}));
var SchemaObjectType;
(function (SchemaObjectType) {
    SchemaObjectType["array"] = "array";
    SchemaObjectType["stringArray"] = "string[]";
    SchemaObjectType["boolean"] = "boolean";
    SchemaObjectType["object"] = "object";
    SchemaObjectType["number"] = "number";
    SchemaObjectType["string"] = "string";
    SchemaObjectType["integer"] = "integer";
    SchemaObjectType["enum"] = "enum";
    SchemaObjectType["null"] = "null";
    SchemaObjectType["union"] = "union";
    SchemaObjectType["file"] = "file";
})(SchemaObjectType || (exports.SchemaObjectType = SchemaObjectType = {}));
var PriorityRule;
(function (PriorityRule) {
    PriorityRule["include"] = "include";
    PriorityRule["exclude"] = "exclude";
    PriorityRule["both"] = "both";
})(PriorityRule || (exports.PriorityRule = PriorityRule = {}));
var ReactQueryMode;
(function (ReactQueryMode) {
    ReactQueryMode["react"] = "react";
    ReactQueryMode["vue"] = "vue";
})(ReactQueryMode || (exports.ReactQueryMode = ReactQueryMode = {}));
const displayReactQueryMode = (mode) => {
    return {
        react: '@tanstack/react-query',
        vue: '@tanstack/vue-query',
    }[mode];
};
exports.displayReactQueryMode = displayReactQueryMode;
