// @uni-helper/unocss-preset-uni@0.2.11/dist/index.mjs
import { presetAttributify, presetUno, definePreset } from 'unocss';
import { isMp, builtInPlatforms, platform } from '@uni-helper/uni-env';
import { presetLegacyCompat } from '@unocss/preset-legacy-compat';
import { presetRemRpx, presetApplet, transformerAttributify } from 'unocss-applet';
import { h } from '@unocss/preset-mini/utils';
import { variantGetParameter } from '@unocss/rule-utils';

function parseOption(value, defaultValue) {
  if (value === false)
    return value;
  if (value === true)
    return defaultValue ?? {};
  if (value)
    return { ...defaultValue, ...value };
  if (defaultValue)
    return defaultValue;
  return false;
}
function resolveOptions(userOptions = {}) {
  const uno = parseOption(userOptions.uno, {});
  const attributify = parseOption(
    userOptions.attributify,
    { ignoreAttributes: isMp ? ["block", "fixed"] : undefined }
  );
  const remRpx = parseOption(
    userOptions.remRpx,
    { mode: isMp ? undefined : "rpx2rem" }
  );
  return {
    ...userOptions,
    uno,
    remRpx,
    attributify
  };
}

function createPresets(options) {
  const presets = [];
  let presetUno$1 = presetUno;
  if (isMp) {
    presetUno$1 = presetApplet;
    presets.push(presetLegacyCompat({
      commaStyleColorFunction: true,
      legacyColorSpace: true
    }));
  }
  if (options.uno)
    presets.push(presetUno$1(options.uno));
  if (options.remRpx)
    presets.push(presetRemRpx(options.remRpx));
  if (options.attributify)
    presets.push(presetAttributify(options.attributify));
  return presets;
}

const theme = {
  platforms: builtInPlatforms.reduce((acc, platform) => {
    acc[platform] = platform;
    const withoutPrefix = platform.replace(/^mp-/, "");
    if (withoutPrefix && withoutPrefix !== platform)
      acc[withoutPrefix] = platform;
    return acc;
  }, { mp: "mp", app: "app", quickapp: "quickapp" })
};

function createTransformers(options) {
  const transformers = [];
  if (isMp && options.attributify)
    transformers.push(transformerAttributify(options.attributify));
  return transformers;
}

function createVariants() {
  const platformVariants = {
    name: "unocss-preset-uni-platforms",
    match(matcher, ctx) {
      const variant = variantGetParameter("uni-", matcher, ctx.generator.config.separators);
      if (variant) {
        const [match, rest] = variant;
        let matchPlatform = h.bracket(match) ?? "";
        const { platforms = {} } = ctx.theme;
        matchPlatform = matchPlatform === "" ? platforms[match] ?? "" : matchPlatform;
        if (matchPlatform) {
          return {
            matcher: rest,
            selector: (s) => platform !== undefined && platform.startsWith(matchPlatform) ? s : `${s}-pass`
          };
        }
      }
    },
    multiPass: true,
    autocomplete: "uni-$platforms:"
  };
  return [platformVariants];
}

const presetUni = definePreset((userOptions = {}) => {
  const options = resolveOptions(userOptions);
  const presets = createPresets(options);
  const variants = createVariants();
  const transformers = createTransformers(options);
  return {
    name: "unocss-preset-uni",
    presets,
    variants,
    theme,
    configResolved(config) {
      if (!config.transformers)
        config.transformers = transformers;
      else
        config.transformers = [...config.transformers, ...transformers];
    }
  };
});

export { createPresets, createTransformers, createVariants, presetUni, resolveOptions, theme };
